### PR TITLE
nix build: add --print-out-paths flag

### PR DESCRIPTION
**Pull request by Artturin**
has the same functionality as default nix-build

$ nix-build . -A "bash" -A "bash.dev" -A "tinycc"
/nix/store/4nmqxajzaf60yjribkgvj5j54x9yvr1r-bash-5.1-p12
/nix/store/c49i1ggnr5cc8gxmk9xm0cn961z104dn-bash-5.1-p12-dev
/nix/store/dbapb08862ajgaax3621fz8hly9fdah3-tcc-0.9.27+date=2022-01-11

$ nix-build . -A "bash"
/nix/store/4nmqxajzaf60yjribkgvj5j54x9yvr1r-bash-5.1-p12

$ $HOME/nixgits/nix/result/bin/nix build "nixpkgs#bash" "nixpkgs#bash.dev" "nixpkgs#tinycc" --print-store-paths
/nix/store/4nmqxajzaf60yjribkgvj5j54x9yvr1r-bash-5.1-p12
/nix/store/c49i1ggnr5cc8gxmk9xm0cn961z104dn-bash-5.1-p12-dev
/nix/store/dbapb08862ajgaax3621fz8hly9fdah3-tcc-0.9.27+date=2022-01-11

$ $HOME/nixgits/nix/result/bin/nix build "nixpkgs#bash" --print-store-paths
/nix/store/4nmqxajzaf60yjribkgvj5j54x9yvr1r-bash-5.1-p12

Supersedes https://github.com/NixOS/nix/pull/2622
Closes https://github.com/NixOS/nix/issues/5601 https://github.com/NixOS/nix/issues/1930

view the first commit for code changes.

@Mic92 @makefu @mickours @domenkozar @fzakaria @tomberek @edolstra 

this was a really useful feature of `nix-build` and may be the reason why scripts haven't switched over to `nix build`
it would be really really useful to have this feature in `nix build`

one of my use cases
```bash
font=$(nix-build '<nixos>' -A liberation_ttf)/share/fonts/truetype/LiberationSans-Regular.ttf
mogrify -font $font -pointsize 50 -fill white -draw 'text 100,100'$COLOR /tmp/tmpcolor.png
```

i wasn't sure where to put the test so i just put it in the remote build test

### CODE REVIEW

File: src/nix/build.cc [hunk 1]
```diff
@@ -82,6 +90,45 @@ struct CmdBuild : InstallablesCommand, MixDryRun, MixJSON, MixProfile
...
+            std::vector<StorePath> result;
+            for (auto & buildable : buildables) {
+                std::visit(overloaded {
+                    [&](const BuiltPath::Opaque & bo) {
+                        result.push_back(bo.path);
+                    },
+                    [&](const BuiltPath::Built & bfd) {
+                        for (auto & output : bfd.outputs) {
+                            result.push_back(output.second);
+                        }
```
[the remark this answers is not in the document]
Comment (Artturin) [reply src/nix/build.cc:120]: i copied the code from https://github.com/NixOS/nix/blob/a61809722fc85a803d412d66667e02e35a8d229a/src/libcmd/command.cc#L242

```
rg  'for.+\{' | wc -l
654
```
:laughing:

[the remark this answers is not in the document]
Comment (Artturin) [reply src/nix/build.cc:120]: it doesn't work here
```
src/nix/build.cc: In lambda function:
src/nix/build.cc:117:25: error: expected ‘{’ before ‘result’
  117 |                         result.push_back(bo.path);
      |                         ^~~~~~
src/nix/build.cc: In member function ‘virtual void CmdBuild::run(nix::ref<nix::Store>)’:
src/nix/build.cc:115:39: error: missing template arguments before ‘{’ token
  115 |                 std::visit(overloaded {
      |                                       ^
make: *** [mk/patterns.mk:3: src/nix/build.o] Error 1
```

**Comment by Artturin** [comment]
updated todo's

**Comment by 06kellyjac** [comment]
yeah, it'd be nice to not have to output json into jq :+1:

**Comment by edolstra** [comment]
From the CLI guidelines:

> When needed, commands should offer a --json flag to allow users to easily parse and script the CLI.

It was a conscious decision not to have ad-hoc, non-extensible, unstructured output on stdout in the new CLI. But I realize it's annoying to have to rely on a tool like jq...

**Comment by 06kellyjac** [comment]
Yeah we probably shouldn't add flags for every output option under the sun, the json output is very useful. But since this was the previous behavior of `nix-build` and came in handy I'd say it's worth it.

Or maybe we could consider having jq style templating build in. Then you can do about anything without a 2nd tool:
`--template '.[].outputs.out'`
like kubectl has a json-template or go-template output format
`kubectl get pods --selector=app=cassandra -o jsonpath='{.items[*].metadata.labels.version}'`
`kubectl get secret my-secret -o go-template='{{range $k,$v := .data}}{{"### "}}{{$k}}{{"\n"}}{{$v|base64decode}}{{"\n\n"}}{{end}}'`

**Comment by Artturin** [comment]
Github cli has a --jq option which doesn't seem to depend on jq being installed because jq isn't in the deps https://archlinux.org/packages/community/x86_64/github-cli/

https://github.com/cli/cli/blob/d794a929da09741f593572c11822b21289cc2f74/pkg/cmdutil/json_flags.go#L90

https://github.com/cli/cli/blob/d794a929da09741f593572c11822b21289cc2f74/pkg/cmdutil/json_flags.go#L140

**Comment by ncfavier** [comment]
I don't think adding JSON filtering to nix is necessary. `--print-store-paths` (or `--print-out-paths`) and `--json` should be enough for most purposes.

**Comment by Artturin** [comment]
I agree

**Comment by 06kellyjac** [comment]
> Github cli has a --jq option which doesn't seem to depend on jq being installed because jq isn't in the deps

It uses a go library that re-implements jq

github.com/itchyny/gojq

**Comment by Artturin** [comment]
added a release note

**Comment by Artturin** [comment]
any code reviews? :slightly_smiling_face:

**Comment by ncfavier** [comment]
Not a code review, but the obligatory bikeshedding over naming: I'd prefer `--print-out-paths`. I think it's more precise, and fits better with `--out-link` and `drv.outPath`. `--print-store-paths` sounds like it could mean "print all evaluated store paths" or "all fetched store paths" or something else.

**Comment by tomberek** [comment]
“ flag to print the resulting store paths”. 

I’m not sure what this means. What is the expectation people can rely on? (This question may also apply to the JSON output.) What will the output contain if combined with remote builders/ eval stores/substitution/symlink generated. What invariant is provided?  Eg: Given identical starting stores where one does the build and the other only gets this stdout - what steps can be taken to reach the same Store state, or similar? Does (build + GC ~= substitute with stdout paths)?

The common use case I presume is as a CLI ad-box variant of a post-build-hook. What are the semantics compared to this pseudo code:

‘’’
nix build —option post-build-hook ‘printf “%s\n” $OUT_PATHS’
‘’’


My intention is not to support or oppose, but to ensure there is consistency so we can apply it in more situations or to understand the distinctions.

**Comment by Artturin** [comment]
@tomberek
```bash
> $HOME/nixgits/nix/result/bin/nix build --option builders 'ssh://eu.nixbuild.net' -j0 nixpkgsMaster#bash --print-output-paths
/nix/store/dndi916j6yxzfzzj2sma2llhrlwahq06-bash-5.1-p16

> $HOME/nixgits/nix/result/bin/nix build --option builders 'ssh://eu.nixbuild.net' -j0 /nix/store/dndi916j6yxzfzzj2sma2llhrlwahq06-bash-5.1-p16 --print-output-paths
/nix/store/dndi916j6yxzfzzj2sma2llhrlwahq06-bash-5.1-p16

> $HOME/nixgits/nix/result/bin/nix build --option builders 'ssh://eu.nixbuild.net' -j0 nixpkgsMaster#bash.dev --print-output-paths
/nix/store/lrrv3fzii19il2lsqdcb5d9mia2wxy5r-bash-5.1-p16-dev



> $HOME/nixgits/nix/result/bin/nix build nixpkgsMaster#bash --print-output-paths
/nix/store/dndi916j6yxzfzzj2sma2llhrlwahq06-bash-5.1-p16

> $HOME/nixgits/nix/result/bin/nix build /nix/store/dndi916j6yxzfzzj2sma2llhrlwahq06-bash-5.1-p16 --print-output-paths
/nix/store/dndi916j6yxzfzzj2sma2llhrlwahq06-bash-5.1-p16

> $HOME/nixgits/nix/result/bin/nix build nixpkgsMaster#bash.dev --print-output-paths
/nix/store/lrrv3fzii19il2lsqdcb5d9mia2wxy5r-bash-5.1-p16-dev
```

**Comment by Artturin** [comment]
there's a difference when the derivation is built for the first time / rebuilt

```bash
nix-build -A bash --check
/nix/store/dndi916j6yxzfzzj2sma2llhrlwahq06-bash-5.1-p16
```

```bash
$HOME/nixgits/nix/result/bin/nix build ".#bash" --print-output-paths --rebuild
/nix/store/lrrv3fzii19il2lsqdcb5d9mia2wxy5r-bash-5.1-p16-dev
/nix/store/s6bspl3x0s8h626vc4ikgv1h8sphlm17-bash-5.1-p16-doc
/nix/store/g0n1a3vdx64zkkprfl0y8643320bsb0a-bash-5.1-p16-info
/nix/store/ddgnym0a3nl1ggz50gg8q1cs0arzic87-bash-5.1-p16-man
/nix/store/dndi916j6yxzfzzj2sma2llhrlwahq06-bash-5.1-p16
```

**Comment by Artturin** [comment]
with `nix build` when a derivation is built for the first time / rebuilt then all the outputs of it will be symlinked to the current dir

while with `nix-build`  only the main output is symlinked to the current dir

**Comment by Artturin** [comment]
i rebuilt my system with
```nix
(self: super: {
  nixUnstable = super.nixUnstable.overrideAttrs (old: {
    src = super.fetchFromGitHub {
      owner = "nixos";
      repo = "nix";
      rev = "abab65f774169c4ebf58823af2a3c327a560deee";
      sha256 = "sha256-kFsJ/WsopJIYKHBpnVGuyXlqdwcrezvvmbT2ycN89r0=";
    };
  });
})
```

i then ran
```
nix build "nixpkgs#bash.dev" --print-output-paths --rebuild
/nix/store/c49i1ggnr5cc8gxmk9xm0cn961z104dn-bash-5.1-p12-dev
```
and
```
nix build "nixpkgs#bash" --print-output-paths --rebuild
/nix/store/4nmqxajzaf60yjribkgvj5j54x9yvr1r-bash-5.1-p12
```

both work

Thank you @edolstra !!

**Review by edolstra — APPROVED** [review]

File: src/nix/build.cc [hunk 2]
```diff
@@ -93,6 +101,28 @@ struct CmdBuild : InstallablesCommand, MixDryRun, MixJSON, MixProfile
                     }, buildable.raw());
                 }
 
+        if (printOutputPaths) {
+            std::vector<StorePath> result;
+            for (auto & buildable : buildables) {
+                std::visit(overloaded {
+                    [&](const BuiltPath::Opaque & bo) {
+                        result.push_back(bo.path);
```
[the remark this answers is not in the document]
Comment (Artturin) [reply src/nix/build.cc]: no reason, i changed it to just print it

**Comment by erikarvstedt** [comment]
I agree with @ncfavier that this should be renamed to `--print-out-paths`, analogous to `--out-link`.

**Comment by edolstra** [comment]
Alternatively, maybe `--out-link` should be renamed to `--output-link`, since there is no good reason why we're abbreviating "output"...

**Comment by bjornfor** [comment]
> Alternatively, maybe `--out-link` should be renamed to `--output-link`, since there is no good reason why we're abbreviating "output"...

The variable inside builders is called `$out` not `$output` though. (Not arguing against rename, just saying.)

**Comment by Artturin** [comment]
print-output-paths sounds good to me

**Comment by Artturin** [comment]
print-out-paths sounds good to me

**Review by tomberek — COMMENTED** [review]
functions and runs as expected. Consistency makes me wonder if any other commands should respect this output format. `nix copy`? `nix profile`?

**Comment by Artturin** [comment]
resolved conflict

**Comment by Artturin** [comment]
conflict resolved.

**Review by thufschmitt — APPROVED** [review]
Thanks, sorry for the long delay :)

I’ve resolved the small conflict. Waiting for https://github.com/NixOS/nix/pull/6437 to be merged so that we can have a green CI, and I’ll merge after that

**Comment by AleXoundOS** [comment]
Sorry, but can it be explained why is `--print-out-paths` not the default behavior? Are there any drawbacks for printing output paths by default for anybody?

**Comment by AleXoundOS** [comment]
After years of quickly digging into just-built out paths, using in scripts, etc and seeing no problem with it. And the number of the out paths is rarely > 5.
It can be really hard to comprehend it, honestly. The only hope is that this will make nix look less "geeky" for newcomers. But the rhetoric question - is nix generally ready to become so user-friendly, that the built out paths information is that much irrelevant to be suppressed. Or a new CLI is going to take credits for a fantastic nix breakthrough :rocket:. Thank you for your work.

**Comment by rockboynton** [comment]
Is there a way to do this via `nix copy` as well?

**Pull request merged**